src/Compilers/CSharp/Portable/Symbols/MemberSymbolExtensions.cs

```diff
@@ -380,44 +380,71 @@ internal static bool HasThisConstructorInitializer(this MethodSymbol method)
+            // parameterless constructor is a synthesized default constructor that is not emitted.
+            static bool isDefaultValueTypeConstructor(NamedTypeSymbol containingType, ConstructorInitializerSyntax initializerSyntax)
+            {
+                if (initializerSyntax.ArgumentList.Arguments.Count > 0)
+                {
+                    return false;
+                }
+                var constructor = containingType.InstanceConstructors.SingleOrDefault(m => m.ParameterCount == 0);
+                return constructor?.IsDefaultValueTypeConstructor(requireNoInitializers: true) == true;
+            }
         }
 
         /// <summary>
-        /// NOTE: every struct has a public parameterless constructor either used-defined or default one
+        /// NOTE: every struct has a public parameterless constructor either user-defined or default one
         /// </summary>
         internal static bool IsParameterlessConstructor(this MethodSymbol method)
         {
             return method.MethodKind == MethodKind.Constructor && method.ParameterCount == 0;
         }
 
         /// <summary>
-        /// default zero-init constructor symbol is added to a struct when it does not define 
-        /// its own parameterless public constructor.
-        /// We do not emit this constructor and do not call it 
+        /// Returns true if the method is the default zero-init constructor synthesized for struct types.
+        /// If the struct type is from metadata, the default constructor is synthesized when there
+        /// is no accessible parameterless constructor. If the struct type is from source, a parameterless
+        /// constructor is synthesized if there is no explicit parameterless constructor, but the synthesized
+        /// constructor is only considered the default zero-init constructor if the synthesized constructor
+        /// would not be emitted. (The default constructor will be emitted if the struct type has field
+        /// initializers and no explicit constructors.)
         /// </summary>
-        internal static bool IsDefaultValueTypeConstructor(this MethodSymbol method)
+        internal static bool IsDefaultValueTypeConstructor(this MethodSymbol method, bool requireNoInitializers = false)
         {
-            return method.IsImplicitlyDeclared &&
-                   method.ContainingType.IsValueType &&
-                   method.IsParameterlessConstructor();
+            if (method.IsImplicitlyDeclared &&
+                method.ContainingType.IsValueType &&
+                method.IsParameterlessConstructor())
+            {
+                if (!requireNoInitializers)
+                {
+                    return true;
+                }
+                var containingType = method.ContainingType.OriginalDefinition;
+                var constructors = containingType.InstanceConstructors;
+                return constructors.Length > 1 || !containingType.HasFieldInitializers();
```

Comment:
>constructors.Length > 1 [](http://example.com/codeflow?start=23&length=23)

Do we need the `constructors.Length > 1` check?

Reply:
> Do we need the `constructors.Length > 1` check?

Yes, I think so, because if there are field initializers and an explicit constructor with parameters, the implicit parameterless constructor is treated as the zero-init constructor (and does not execute field initializers).

Reply:
Looking at this again, I stumbled on it again. Consider leaving a comment